src/MyConnection.h

```diff
@@ -30,15 +35,21 @@ class MyConnection : boost::noncopyable {
     mysql_options(pConn_, MYSQL_OPT_LOCAL_INFILE, 0);
     // Default to UTF-8
     mysql_options(pConn_, MYSQL_SET_CHARSET_NAME, "UTF8");
-    if (groups != "")
-      mysql_options(pConn_, MYSQL_READ_DEFAULT_GROUP, groups.c_str());
-    if (default_file != "")
-      mysql_options(pConn_, MYSQL_READ_DEFAULT_FILE, default_file.c_str());
-
-
-    if (!mysql_real_connect(pConn_, host.c_str(), user.c_str(),
-        password.c_str(), db == "" ? NULL : db.c_str(), port,
-        unix_socket == "" ? NULL : unix_socket.c_str(), client_flag)) {
+    if (!groups.isNull())
+      mysql_options(pConn_, MYSQL_READ_DEFAULT_GROUP,
+                    Rcpp::as<std::string>(groups).c_str());
```

Comment:
How about:

```C++
class as_cstr {
public:
  as_cstr(const Rcpp::Nullable<Rcpp::CharacterVector>& val) :
    val_(val), buffer_(Rcpp::as<std::string>(val)) {}

  operator const char*() const {
    if (val_.isNull()) {
      return NULL;
    }
    else {
      return buffer_.c_str()
    }
  }

private:
  const Rcpp::Nullable<Rcpp::CharacterVector>& val_;
  std::string buffer_;
}
```

True, it's not a function anymore, but pretty much function-like. Perhaps C++11 can offer neater alternatives.

(Sorry, seems I've really forgotten how troublesome callee-allocated data can be in C++...)